ulib/FStar.Algebra.CommMonoid.Fold.fst

```diff
@@ -0,0 +1,129 @@
+(*
+   Copyright 2008-2018 Microsoft Research
+   
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+   
+       http://www.apache.org/licenses/LICENSE-2.0
+       
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+   Author: A. Rozanov
+*)
+
+module FStar.Algebra.CommMonoid.Fold
+
+open FStar.Seq.Base
+open FStar.Seq.Properties
+open FStar.Seq.Permutation
+
+(* Here we define the notion for big sums and big products for 
+   arbitrary commutative monoids. We construct the folds from 
+   an integer range and a function, then calculate the fold --
+   a sum or a product, depending on the monoid operation. *)
+ 
+(* We refine multiplication a bit to make proofs smoothier *)
+let ( *) (x y: int) : (t:int{(x>0 /\ y>0) ==> t>0}) = op_Multiply x y
+ 
+type under (k: pos) = x:nat{x<k}
+type not_less_than (x: int) = (t: int{t>=x})
+type inbetween (x: int) (y: not_less_than x) = (t: int{t>=x && t<=y})
+type counter_of_range (x: int) (y: not_less_than x) = (t: nat{t<(y+1-x)})
+
+let range_count (x: int) (y: not_less_than x) : pos = (y+1)-x
```

Comment:
These definitions are similar to the ones in a previous PR ... can we put them somewhere common and re-use them?

For example, can we add a module like FStar.Intervals or something like that?

As a side-remark: What's a bit surprising about `inbetween` is that we almost always work with half-open intervals, but this one defines a closed interval.

Reply:
yep, we certainly should. Perhaps add both closed and half-closed versions too. I already noticed this appears occasionally in several ulib files. Perhaps everyone eventually comes up with his own range type :)